File: apps/backend-java/src/main/java/com/enterpriseagentos/backendjava/domain/features/agents/interfaces/IChannelRepository.java
package com.enterpriseagentos.backendjava.domain.features.agents.interfaces;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

import com.enterpriseagentos.backendjava.domain.features.agents.models.AgentChannelBindingModel;
import com.enterpriseagentos.backendjava.domain.features.agents.models.ChannelConnectionModel;

public interface IChannelRepository {
    CompletableFuture<List<ChannelConnectionModel>> listConnectionsAsync();
    CompletableFuture<ChannelConnectionModel> getConnectionAsync(UUID id);
    CompletableFuture<ChannelConnectionModel> createConnectionAsync(ChannelConnectionModel model);
    CompletableFuture<ChannelConnectionModel> updateConnectionAsync(UUID id, java.util.function.Consumer<ChannelConnectionModel> apply);
    CompletableFuture<Boolean> deleteConnectionAsync(UUID id);
    CompletableFuture<List<AgentChannelBindingModel>> listBindingsAsync(UUID agentId);
    CompletableFuture<AgentChannelBindingModel> getBindingAsync(UUID bindingId);
    CompletableFuture<AgentChannelBindingModel> createBindingAsync(AgentChannelBindingModel model);
    CompletableFuture<AgentChannelBindingModel> updateBindingAsync(UUID bindingId, java.util.function.Consumer<AgentChannelBindingModel> apply);
    CompletableFuture<Boolean> deleteBindingAsync(UUID bindingId);
    CompletableFuture<List<AgentChannelBindingModel>> findBindingsByConnectionAsync(UUID connectionId);
    CompletableFuture<List<ChannelConnectionModel>> findConnectionsByChannelTypeAsync(String channelType);
}

File: apps/backend/src/Features/Channels/Domain/IChannelRepository.cs
namespace EnterpriseAgentOs.Domain.Features.Channels;

public interface IChannelRepository
{
    // ---------- Channel Connections ----------
    Task<IReadOnlyList<ChannelConnectionRecord>> ListConnectionsAsync(ChannelConnectionFilter? filter = null, CancellationToken ct = default);
    Task<ChannelConnectionRecord?> GetConnectionByAsync(ChannelConnectionFilter filter, CancellationToken ct = default);
    Task<ChannelConnectionRecord> CreateConnectionAsync(ChannelConnectionRecord record, CancellationToken ct = default);
    Task<ChannelConnectionRecord?> UpdateConnectionAsync(Guid id, Action<ChannelConnectionRecord> apply, CancellationToken ct = default);
    Task<bool> DeleteConnectionAsync(Guid id, CancellationToken ct = default);

    // ---------- Agent Channel Bindings ----------
    Task<IReadOnlyList<AgentChannelBindingRecord>> ListBindingsAsync(Guid agentId, CancellationToken ct = default);
    Task<AgentChannelBindingRecord?> GetBindingByAsync(AgentChannelBindingFilter filter, CancellationToken ct = default);
    Task<AgentChannelBindingRecord> CreateBindingAsync(AgentChannelBindingRecord record, CancellationToken ct = default);
    Task<AgentChannelBindingRecord?> UpdateBindingAsync(Guid bindingId, Action<AgentChannelBindingRecord> apply, CancellationToken ct = default);
    Task<bool> DeleteBindingAsync(Guid bindingId, CancellationToken ct = default);

    // ---------- Routing queries ----------
    Task<IReadOnlyList<AgentChannelBindingRecord>> FindBindingsByConnectionAsync(Guid connectionId, CancellationToken ct = default);
    Task<IReadOnlyList<ChannelConnectionRecord>> FindConnectionsByChannelTypeAsync(string channelType, CancellationToken ct = default);
}
